tests/Unit/Core/Form/IdentifiableObject/Handler/FormHandlerTest.php

```diff
@@ -0,0 +1,141 @@
+        $formFactoryMock = $this->createMock(FormFactoryInterface::class);
+        $namedFormBuilderMock = $this->createMock(FormBuilderInterface::class);
+        $hookDispatcherMock = $this->createMock(HookDispatcherInterface::class);
+        $dataProviderMock = $this->createMock(FormDataProviderInterface::class);
+
+        // mocks behavior configuration
+        $formBuilderMock
+            ->method('getFormFactory')
+            ->will($this->returnValue($formFactoryMock));
+        $formFactoryMock
+            ->method('createNamedBuilder')
+            ->with($this->equalTo('formA'))
+            ->will($this->returnValue($namedFormBuilderMock));
+
+        $this->handler = new FormHandler(
+            $formBuilderMock,
+            $hookDispatcherMock,
+            $dataProviderMock,
+            ['a' => 'a', 'b' => 'b', 'c' => 'c'],
+            'AHook',
+            'formA'
+        );
+
+        $this->formBuilderMock = $namedFormBuilderMock;
+        $this->hookDispatcherMock = $hookDispatcherMock;
+        $this->dataProviderMock = $dataProviderMock;
+    }
+
+    public function testCanBeConstructed()
+    {
+        $this->assertInstanceOf(FormHandlerInterface::class, $this->handler);
+    }
+
+    public function testGetForm()
+    {
+        $this->formBuilderMock->expects($this->exactly(3))
+            ->method('add')
+            ->withConsecutive(
+                ['a', 'a'],
+                ['b', 'b'],
+                ['c', 'c']
+            );
+
+        $this->dataProviderMock
+            ->method('getData')
+            ->will($this->returnValue(['d' => 'd']));
+
+        $this->formBuilderMock->expects($this->once())
+            ->method('setData')
+            ->with(
+                $this->equalTo(['d' => 'd'])
+            );
+
+        $this->hookDispatcherMock->expects($this->once())
+            ->method('dispatchWithParameters')
+            ->with(
+                $this->equalTo('actionAHookForm'),
+                $this->equalTo(['form_builder' => $this->formBuilderMock])
+            );
+
+        $formMock = $this->createMock(\Symfony\Component\Form\FormInterface::class);
+        $this->formBuilderMock
+            ->method('getForm')
+            ->will($this->returnValue($formMock));
+
+        $form = $this->handler->getForm();
+
+        $this->assertEquals($formMock, $form);
+    }
+
+    public function testSaveData()
+    {
+        $this->dataProviderMock
+            ->method('setData')
+            ->with($this->equalTo(['x' => 'y']))
+            ->will($this->returnValue(['err' => 'or']));
+
+        $this->hookDispatcherMock->expects($this->once())
+            ->method('dispatchWithParameters')
+            ->with(
+                $this->equalTo('actionAHookSave'),
+                $this->equalTo([
+                    'errors' => ['err' => 'or'],
+                    'form_data' => ['x' => 'y'],
+                ])
+            );
+
+        $this->handler->save(['x' => 'y']);
```

Comment:
Didn't you forget an `assert` statement?

Reply:
The return of `$this->handler->save(...)` is an array `$errors` directly taken from the DataProvider which is a mock here. So I thought it was not relevant to check, but I can add it.

https://github.com/PrestaShop/PrestaShop/blob/develop/src/Core/Form/FormHandler.php#L123
```
public function save(array $data)
    {
        $errors = $this->formDataProvider->setData($data);
        $this->hookDispatcher->dispatchWithParameters(
            "action{$this->hookName}Save",
            [
                'errors' => &$errors,
                'form_data' => &$data,
            ]
        );
        return $errors;
    }
```

Reply:
Nope, that's fine by me.